tket2/src/serialize/pytket/encoder.rs

```diff
@@ -169,21 +188,345 @@ impl JsonEncoder {
             commands: self.commands,
-            qubits: self.qubit_registers,
-            bits: self.bit_registers,
-            implicit_permutation: self.implicit_permutation,
+            qubits,
+            bits,
+            implicit_permutation,
+        }
+    }
+
+    /// Translate a linear [`CircuitUnit`] into a [`RegisterUnit`], if possible.
+    fn unit_to_register(&self, unit: CircuitUnit) -> Option<RegisterUnit> {
+        match unit {
+            CircuitUnit::Linear(i) => self.qubits.get(i).cloned(),
+            CircuitUnit::Wire(wire) => self.bits.get(&wire).cloned(),
+        }
+    }
+}
+
+/// A structure for tracking qubits used in the circuit being encoded.
+///
+/// Nb: Although `tket-json-rs` has a "Register" struct, it's actually
+/// an identifier for single qubits in the `Register::0` register.
+/// We rename it to `RegisterUnit` here to avoid confusion.
+#[derive(Debug, Clone, Default)]
+struct QubitTracker {
+    /// The ordered TKET1 names for the input qubit registers.
+    inputs: Vec<RegisterUnit>,
+    /// The ordered TKET1 names for the output qubit registers.
+    outputs: Option<Vec<RegisterUnit>>,
+    /// The TKET1 qubit registers associated to each qubit unit of the circuit.
+    qubit_to_reg: HashMap<usize, RegisterUnit>,
+}
+
+impl QubitTracker {
+    /// Create a new [`QubitTracker`] from the bit inputs of a [`Circuit`].
+    /// Reads the [`METADATA_Q_REGISTERS`] metadata entry with preset pytket bit register names.
+    ///
+    /// If the circuit contains more bit inputs than the provided list,
+    /// new registers are created for the remaining bits.
+    pub fn new(circ: &Circuit<impl HugrView>) -> Self {
+        let mut tracker = QubitTracker::default();
+
+        if let Some(input_regs) = circ
+            .hugr()
+            .get_metadata(circ.parent(), METADATA_Q_REGISTERS)
+        {
+            tracker.inputs = serde_json::from_value(input_regs.clone()).unwrap();
+        }
+        let output_regs = circ
+            .hugr()
+            .get_metadata(circ.parent(), METADATA_Q_OUTPUT_REGISTERS)
+            .map(|regs| serde_json::from_value(regs.clone()).unwrap());
+        if let Some(output_regs) = output_regs {
+            tracker.outputs = Some(output_regs);
+        }
+
+        let qubit_count = circ.units().filter(|(_, _, ty)| ty == &QB_T).count();
+
+        for i in 0..qubit_count {
+            // Use the given input register names if available, or create new ones.
+            if let Some(reg) = tracker.inputs.get(i) {
+                tracker.qubit_to_reg.insert(i, reg.clone());
+            } else {
+                let reg = tracker.add_qubit_register(i).clone();
+                tracker.inputs.push(reg);
+            }
+        }
+
+        tracker
+    }
+
+    /// Add a new register unit for a bit wire.
+    pub fn add_qubit_register(&mut self, unit_id: usize) -> &RegisterUnit {
+        let reg = RegisterUnit("q".to_string(), vec![self.qubit_to_reg.len() as i64]);
```

Comment:
if "q" register isn't already being used this could suddenly create a very large "q" register with mostly unused qubits right? Not too much of a problem for tket but if then lowered to something that creates whole registers would be annoying. Maybe worth maintaining a "next free qubit"?